### PR TITLE
Gemma 2

### PR DESCRIPTION
**Pull request by talumbau**
Support Gemma 2
    
     -  add support for local sliding window attention
     -  add config for pre- and post- Feed forward layer norm
     -  add config for final logit softcap
     -  SDPA with logit softcap (HLFB not supported yet)
     -  Gemma2Block for immediate post-attention norm
     -  ModelConfig for Gemma 2
     -  Matches goldens from Gemma PyTorch implementation.

### CODE REVIEW

File: ai_edge_torch/generative/examples/gemma/gemma.py [hunk 1]
```diff
@@ -36,16 +37,73 @@
...
     attn_output_proj="model.layers.{}.self_attn.o_proj",
     pre_attn_norm="model.layers.{}.input_layernorm",
-    pre_ff_norm="model.layers.{}.post_attention_layernorm",
+    post_attn_norm="model.layers.{}.post_attention_layernorm",
     embedding="model.embed_tokens",
     final_norm="model.norm",
     lm_head=None,
 )
 
+TENSOR_NAMES_GEMMA2 = loading_utils.ModelLoader.TensorNames(
```
Comment (haozha111) [inline ai_edge_torch/generative/examples/gemma/gemma.py:46]: Can we keep gemma2 related code in a new file? this might be easier for future development.

File: ai_edge_torch/generative/examples/gemma/gemma.py [hunk 2]
```diff
@@ -36,16 +37,73 @@
...
+    return output
+
+
+T = TypeVar("T", TransformerBlock, Gemma2Block)
+
 
 class Gemma(nn.Module):
 
-  def __init__(self, config: cfg.ModelConfig):
+  def __init__(self, config: cfg.ModelConfig, block_type: T = TransformerBlock):
```
Comment (haozha111) [inline ai_edge_torch/generative/examples/gemma/gemma.py:106]: oh i see, since you are reusing the logic here, it might actually save some code.

File: ai_edge_torch/generative/examples/gemma/gemma.py (hunk 1 above)
Comment (hheydary) [reply ai_edge_torch/generative/examples/gemma/gemma.py:46]: +1 on separating Gemma2 to another file. 

I know that would not be DRY but a little bit of repetition should be ok if it make the code easier to follow. At the end of the day, most of the common blocks are moved out to the layers directory.

File: ai_edge_torch/generative/examples/stable_diffusion/diffusion.py [hunk 3]
```diff
@@ -250,6 +250,7 @@ def __init__(self, config: unet_cfg.DiffusionModelConfig):
 
     attention_config = layers_cfg.AttentionConfig(
         num_heads=config.transformer_num_attention_heads,
+        head_dim=block_out_channels[0] // config.transformer_num_attention_heads,
```
Comment (yichunk) [inline ai_edge_torch/generative/examples/stable_diffusion/diffusion.py:253]: Please refer to https://github.com/google-ai-edge/ai-edge-torch/pull/87/files#r1678556679, if #87 is not going to be submitted.
Same for the configs in stable_diffusion_loader.

**Review by yichunk — CHANGES_REQUESTED** [review]

File: ai_edge_torch/generative/examples/gemma/gemma.py [hunk 4]
```diff
@@ -78,11 +136,26 @@ def __init__(self, config: cfg.ModelConfig):
...
+      )
+    else:
+      self.sliding_window_mask_cache = None
+
     self.config = config
 
-  # The model's forward function takes in additional k/v cache tensors
-  # and returns the updated k/v cache tensors to the caller.
-  # This can be eliminated if we handle k/v cache updates inside the model itself.
+  @torch.inference_mode
```
[the remark this answers is not in the document]
Comment (haozha111) [reply ai_edge_torch/generative/examples/gemma/gemma.py:151]: I think torch.inference_modes is still required if we want to export an inference graph, and that will apply a context manager which will remove all the gradient ops from the graph. IMO this is needed for best performance..

Maybe we can investigate a bit about why there are warnings thrown from our converter.

[the remark this answers is not in the document]
Comment (haozha111) [reply ai_edge_torch/generative/examples/gemma/gemma.py:151]: https://pytorch.org/docs/stable/generated/torch.autograd.grad_mode.inference_mode.html

File: ai_edge_torch/generative/layers/model_config.py [hunk 5]
```diff
@@ -123,8 +139,10 @@ class ModelConfig:
   pre_attention_norm_config: NormalizationConfig = field(
```
[the remark this answers is not in the document]
Comment (haozha111) [reply ai_edge_torch/generative/layers/model_config.py:139]: +1 to keep the configurations as local as possible..

File: ai_edge_torch/generative/examples/gemma/gemma.py (hunk 1 above)
Comment (talumbau) [reply ai_edge_torch/generative/examples/gemma/gemma.py]: Moved Gemma2 to a new file. PTAL.

File: ai_edge_torch/generative/examples/gemma/gemma.py (hunk 2 above)
Comment (talumbau) [reply ai_edge_torch/generative/examples/gemma/gemma.py]: Yes, I thought it would be reasonable to just config for Gemma2, but it's fine as a new file

File: ai_edge_torch/generative/layers/model_config.py [hunk 6]
```diff
@@ -53,11 +53,17 @@ class FeedForwardType(enum.Enum):
   GATED = enum.auto()
 
 
+class AttentionType(enum.Enum):
+  GLOBAL = 1
```
[the remark this answers is not in the document]
Comment (talumbau) [reply ai_edge_torch/generative/layers/model_config.py]: Done.

File: ai_edge_torch/generative/examples/stable_diffusion/diffusion.py (hunk 3 above)
Comment (talumbau) [reply ai_edge_torch/generative/examples/stable_diffusion/diffusion.py]: #87 is now closed in favor of #120 which this PR is based on.

File: ai_edge_torch/generative/examples/gemma/gemma.py [hunk 7]
```diff
@@ -142,23 +219,77 @@ def get_model_config_2b(kv_cache_max_len: int = 1024) -> cfg.ModelConfig:
...
   return config
 
 
+def build_v2_2b_model(checkpoint_path, **kwargs) -> nn.Module:
+  config = get_model_config_v2_2b(**kwargs)
+  model = Gemma(config, block_type=Gemma2Block)
+  loader = loading_utils.ModelLoader(checkpoint_path, TENSOR_NAMES_GEMMA2)
+  # since embedding and lm-head use the same weight, we need to set strict
+  # to False.
+  loader.load(model, strict=False)
```
[the remark this answers is not in the document]
Comment (talumbau) [reply ai_edge_torch/generative/examples/gemma/gemma.py]: Good call. Done.

File: ai_edge_torch/generative/layers/model_config.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (talumbau) [reply ai_edge_torch/generative/layers/model_config.py:143]: I filed https://buganizer.corp.google.com/issues/357149907 against myself for this. There's already a number of moving parts here:

 - This PR is based on the (still unlanded) #120 which is a large global change for config'ing head dimension
 - The first commit of this PR changes the name of the pre-feedforward layer norm to "post attention" as prep work for Gemma 2 *actual* pre-feedforward layer norm
 - The second commit of this PR re-introduces the "pre-feedforward layer norm" name and uses it directly prior to applying the feedforward layer

Let's try to get all of that in before making the above change.

**Review by yichunk — APPROVED** [review]

File: ai_edge_torch/generative/examples/gemma/gemma.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (talumbau) [reply ai_edge_torch/generative/examples/gemma/gemma.py]: I think we should make such a change holistically across the repo (inference mode or not). `model.eval()` sounds good for now and I have done that for this model.

**Comment by haozha111** [comment]
can we add a test in https://github.com/google-ai-edge/ai-edge-torch/blob/main/ai_edge_torch/generative/test/test_model_conversion.py ?

**Comment by talumbau** [comment]
> can we add a test in https://github.com/google-ai-edge/ai-edge-torch/blob/main/ai_edge_torch/generative/test/test_model_conversion.py ?

See the new test in `test_model_conversion.py`. Two notes:

- the test passes when you comment out the `skipTest`. I skip the test since that it what is done with Gemma for right now
- in order to get the "fake" model passed the converter, I needed to add back in the `@torch.inference_mode` decorator to the `forward` method the main model. This is what is done in Gemma. Regarding @hheydary 's comments on those decorators, it seems reasonable to remove them if we can figure out how to get a proper tracing done without. I think we might just want to set a global inference mode before tracing with torchXLA. Anyway, something to handle outside this PR.

**Review by haozha111 — APPROVED** [review]

**Review by majiddadashi — APPROVED** [review]
